xen/arch/arm/vscmi.c

```diff
@@ -0,0 +1,356 @@
+        break;
+    }
+    case PERF_DOMAIN_ATTRIBUTES:
+    {
+        uint32_t domain_id = le32_to_cpu(*(__le32*)data->msg_payload);
+        struct scmi_msg_resp_perf_domain_attributes* attrs =
+            (struct scmi_msg_resp_perf_domain_attributes*)(data->msg_payload + 4);
+
+        if ( domain_id >= current->domain->max_vcpus )
+        {
+            write_32(data->msg_payload, SCMI_ERR_RANGE);
+            data->length = sizeof(uint32_t) * 2;
+            break;
+        }
+        write_32(data->msg_payload, SCMI_SUCCESS);
+
+        attrs->flags = cpu_to_le32(BIT(31) | BIT(30));
+        attrs->rate_limit_us = le32_to_cpu(0);
+        attrs->sustained_freq_khz = le32_to_cpu(PERF_SUSTAINED_FREQ_KHZ);
+        attrs->sustained_perf_level = le32_to_cpu(PERF_OPP_COUNT / 2);
+        snprintf((char*)attrs->name, sizeof(attrs->name), "vcpu%d", domain_id);
+        data->length = sizeof(*attrs) + sizeof(uint32_t) * 2;
+
+        break;
+    }
+    case PERF_DESCRIBE_LEVELS:
+    {
+        struct scmi_msg_perf_describe_levels *req = (void*)data->msg_payload;
+        struct scmi_msg_resp_perf_describe_levels *resp =
+            (void*)(data->msg_payload + 4);
+        int idx = le32_to_cpu(req->level_index);
+        int i;
+
+        if ( le32_to_cpu(req->domain) >= current->domain->max_vcpus ) {
+            write_32(data->msg_payload, SCMI_ERR_RANGE);
+            data->length = sizeof(uint32_t) * 2;
+            break;
+        }
+
+        if ( idx >= PERF_OPP_COUNT ) {
+            write_32(data->msg_payload, SCMI_ERR_RANGE);
+            data->length = sizeof(uint32_t) * 2;
+            break;
+        }
+
+        resp->num_remaining = MAX(0, PERF_OPP_COUNT - idx - PERF_OPPS_PER_CALL);
+        resp->num_returned = MIN(PERF_OPPS_PER_CALL, PERF_OPP_COUNT - idx);
+
+        for ( i = 0; i < resp->num_returned; i++ ) {
+            resp->opp[i].perf_val = idx + i + 1;
+            resp->opp[i].power = resp->opp[i].perf_val;
+            resp->opp[i].transition_latency_us = cpu_to_le16(1);
+        }
+        data->length = sizeof(*resp) + sizeof(uint32_t) * 2 +
+            sizeof(resp->opp[0]) * le16_to_cpu(resp->num_returned);
+
+        write_32(data->msg_payload, SCMI_SUCCESS);
+
+        break;
+    }
+    case PERF_LEVEL_SET:
+    {
+        struct scmi_perf_set_level *req = (void*)data->msg_payload;
+        int domain = le32_to_cpu(req->domain);
+        int level = le32_to_cpu(req->level) - 1;
+
+        if ( domain >= current->domain->max_vcpus ) {
+            write_32(data->msg_payload, SCMI_ERR_RANGE);
+            data->length = sizeof(uint32_t) * 2;
+            break;
+        }
+
+        if ( level < 0 || level >= PERF_OPP_COUNT) {
+            gprintk(XENLOG_WARNING, "vscmi: requested opp is out of bounds: %d\n", level);
+            write_32(data->msg_payload, SCMI_ERR_PARAMS);
+            data->length = sizeof(uint32_t) * 2;
+            break;
+        }
+
```

Comment:
Do we need to take an action if we are already at the required perf level?

if (current->domain->vcpu[domain]->arch.opp == level)